File: 83. Remove Duplicates from Sorted List.cpp
/*
Given a sorted linked list, delete all duplicates such that each element appear only once.

For example,
Given 1->1->2, return 1->2.
Given 1->1->2->3->3, return 1->2->3.
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;
        for (ListNode *first = head, *second = head->next; 
            second != nullptr; second = second->next) {
                if (first->val != second->val) {
                    first->next = second;
                    first = second;
                }
                if (second->next == nullptr)
                    first->next = nullptr;
        }
        return head;
    }
};

File: Array/83. Remove Duplicates from Sorted List.py
# Given a sorted linked list, delete all duplicates such that each element appear only once.

# For example,
# Given 1->1->2, return 1->2.
# Given 1->1->2->3->3, return 1->2->3.

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        dummy = left = ListNode(None)
        prev, cur = head, head.next
        while cur:
            if cur.val != prev.val:
                left.next = prev
                left = left.next
            cur = cur.next
            prev = prev.next
        left.next = prev
        return dummy.next
        